crates/aquascope/src/analysis/permissions/flow.rs

```diff
@@ -0,0 +1,501 @@
+      {
+        FlowEdgeKind::MissingUniversalConstraint
+      }
+
+      // Case 3: flowing from a strictly concrete region.
+      // Local region to abstract sink is disallowed.
+      (AbstractRegionKind::None, AbstractRegionKind::Source(_)) => {
+        FlowEdgeKind::LocalOutlivesUniversal
+      }
+
+      // Case 4: if the region data is flowing from could contain a dangling
+      // pointer we report this as a flow violation (though, I'm not 100%
+      // confident here).
+      (_, _) => {
+        let from = self.constraint_graph.scc(from);
+        for local in self.contains_local.iter(from) {
+          if self.dangling_local_sources.contains(local) {
+            return FlowEdgeKind::LocalInvalidatedAtExit;
+          }
+        }
+
+        FlowEdgeKind::Ok
+      }
+    }
+  }
+}
+
+pub fn compute_flows(ctxt: &mut PermissionsCtxt) {
+  let tcx = ctxt.tcx;
+  let body = &ctxt.body_with_facts.body;
+
+  macro_rules! flat_nodes {
+    ($tups:expr) => {
+      $tups
+        .iter()
+        .flat_map(|&(o1, o2)| SmallVec::<[Origin; 2]>::from_buf([o1, o2]))
+        .unique()
+    };
+  }
+
+  macro_rules! node_count {
+    ($tups:expr) => {
+      flat_nodes!($tups)
+        .minmax_by_key(|v| v.index())
+        .into_option()
+        .map(|(_, mx)| mx.index() + 1)
+        .unwrap_or(0)
+    };
+  }
+
+  // Compute the constraint graph with all regions.
+  let constraints = ctxt
+    .polonius_input_facts
+    .subset_base
+    .iter()
+    .map(|&(o1, o2, _)| (o1, o2))
+    .collect::<Vec<_>>();
+
+  let vertices = flat_nodes!(constraints).collect::<Vec<_>>();
+
+  // NOTE: regions that only occur in the return type are not
+  // included in the abstract placeholders set. Example:
+  // ```rust,no-run
+  // fn mk_string() -> &'a String {
+  //   let s = String::from("s");
+  //   &s
+  // }
+  // ```
+  // The `'a`, included for clarity but would be implicit usually,
+  // does not appear in the placeholders set.
+  let placeholders = ctxt
+    .polonius_input_facts
+    .placeholder
+    .iter()
+    .filter_map(|&(p, _)| vertices.contains(&p).then_some(p))
+    .chain(body.regions_in_return().map(|rg| rg.to_region_vid()))
+    .collect::<Vec<_>>();
+
+  log::debug!("PLACEHOLDERS: {placeholders:#?}");
+
+  let placeholder_edges = ctxt
+    .polonius_input_facts
+    .known_placeholder_subset
+    .iter()
+    .filter(|(f, t)| vertices.contains(f) && vertices.contains(t))
+    .copied()
+    .chain(placeholders.iter().map(|&o| (o, o)))
+    .collect::<Vec<_>>();
+
+  // Graph of constraints that need to be satisfied. This shows
+  // us how data flows from one region into another.
+  let constraint_graph = VecGraph::new(node_count!(constraints), constraints);
+
+  // Allowed flows between abstract regions.
+  let known_flows_graph =
+    VecGraph::new(node_count!(placeholder_edges), placeholder_edges);
+
+  let scc_constraints = Sccs::<Origin, SccIdx>::new(&constraint_graph);
```

Comment:
Do we need to compute the Sccs when we already have the `subset_anywhere` relation in the Polonius output?

Reply:
`subset_anywhere` is only computed when the location-insensitive version of the Polonius algorithm is used. Since we're using the `Naive` variant, the map isn't filled. [https://github.com/rust-lang/polonius/blob/master/polonius-engine/src/output/mod.rs#L285-L309]